File: scripts/gateway_supervisor.py
"""Discover, start, and stop the memory-tencentdb Gateway Node.js sidecar.

Search order for `src/gateway/server.ts` mirrors the upstream Hermes
supervisor exactly so an existing upstream checkout is reused as-is:

  1. MEMORY_TENCENTDB_GATEWAY_CMD env var (full command override)
  2. ~/.memory-tencentdb/tdai-memory-openclaw-plugin/src/gateway/server.ts  (preferred)
  3. ~/tdai-memory-openclaw-plugin/src/gateway/server.ts                   (legacy)
  4. ~/.hermes/plugins/tdai-memory-openclaw-plugin/src/gateway/server.ts   (hermes-style)

Usage:
    python gateway_supervisor.py start    # spawn + wait for /health
    python gateway_supervisor.py stop     # SIGTERM the supervised PID
    python gateway_supervisor.py status   # print health + pid
"""

from __future__ import annotations

import json
import os
import shlex
import signal
import subprocess
import sys
import time
from pathlib import Path

from gateway_client import GatewayClient, DEFAULT_HOST, DEFAULT_PORT

STATE_DIR = Path.home() / ".memory-tencentdb"
PID_FILE = STATE_DIR / "gateway.pid"
LOG_DIR = STATE_DIR / "logs"
STDOUT_LOG = LOG_DIR / "gateway.stdout.log"
STDERR_LOG = LOG_DIR / "gateway.stderr.log"

HEALTH_INTERVAL = 0.5
HEALTH_MAX_WAIT = 30


def discover_server_ts() -> str | None:
    home = Path.home()
    candidates = [
        home / ".memory-tencentdb" / "tdai-memory-openclaw-plugin" / "src" / "gateway" / "server.ts",
        home / "tdai-memory-openclaw-plugin" / "src" / "gateway" / "server.ts",
        home / ".hermes" / "plugins" / "tdai-memory-openclaw-plugin" / "src" / "gateway" / "server.ts",
    ]
    for c in candidates:
        if c.is_file():
            return str(c)
    return None


def resolve_command() -> str | None:
    env_cmd = os.environ.get("MEMORY_TENCENTDB_GATEWAY_CMD", "").strip()
    if env_cmd:
        return env_cmd
    server_ts = discover_server_ts()
    if server_ts:
        return f"node --import tsx {shlex.quote(server_ts)}"
    return None


def read_pid() -> int | None:
    try:
        return int(PID_FILE.read_text("utf-8").strip())
    except Exception:
        return None


def pid_alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


def start() -> int:
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    LOG_DIR.mkdir(parents=True, exist_ok=True)

    client = GatewayClient()
    try:
        h = client.health(timeout=2)
        print(f"gateway already healthy: {h}")
        return 0
    except Exception:
        pass

    cmd = resolve_command()
    if not cmd:
        print(
            "ERROR: cannot locate Gateway. Either set MEMORY_TENCENTDB_GATEWAY_CMD or run "
            "`/memory-init` to clone the upstream repo.",
            file=sys.stderr,
        )
        return 2

    stdout_f = STDOUT_LOG.open("ab", buffering=0)
    stderr_f = STDERR_LOG.open("ab", buffering=0)

    creationflags = 0
    if os.name == "nt":
        creationflags = getattr(subprocess, "DETACHED_PROCESS", 0) | getattr(
            subprocess, "CREATE_NEW_PROCESS_GROUP", 0
        )

    proc = subprocess.Popen(
        shlex.split(cmd, posix=os.name != "nt"),
        stdout=stdout_f,
        stderr=stderr_f,
        stdin=subprocess.DEVNULL,
        env={**os.environ},
        creationflags=creationflags if os.name == "nt" else 0,
        close_fds=True,
    )
    PID_FILE.write_text(str(proc.pid), "utf-8")
    print(f"spawned gateway pid={proc.pid}, waiting for /health...")

    deadline = time.time() + HEALTH_MAX_WAIT
    while time.time() < deadline:
        if proc.poll() is not None:
            print("ERROR: gateway exited during startup. Tail of stderr:", file=sys.stderr)
            try:
                tail = STDERR_LOG.read_bytes()[-2048:].decode("utf-8", "replace")
                print(tail, file=sys.stderr)
            except Exception:
                pass
            return 3
        try:
            h = client.health(timeout=2)
            status = h.get("status")
            if status in ("ok", "degraded"):
                print(f"gateway ready: {h}")
                return 0
        except Exception:
            pass
        time.sleep(HEALTH_INTERVAL)

    print("ERROR: gateway failed to become healthy within timeout", file=sys.stderr)
    return 4


def stop() -> int:
    pid = read_pid()
    if not pid:
        print("no pid file; nothing to stop")
        return 0
    if not pid_alive(pid):
        print(f"pid {pid} already exited")
        PID_FILE.unlink(missing_ok=True)
        return 0
    try:
        if os.name == "nt":
            subprocess.run(["taskkill", "/PID", str(pid), "/T", "/F"], check=False)
        else:
            os.kill(pid, signal.SIGTERM)
        print(f"sent stop signal to pid {pid}")
    except Exception as e:
        print(f"failed to signal pid {pid}: {e}", file=sys.stderr)
        return 1
    PID_FILE.unlink(missing_ok=True)
    return 0


def status() -> int:
    pid = read_pid()
    client = GatewayClient()
    health = None
    try:
        health = client.health(timeout=2)
    except Exception as e:
        health = {"status": "down", "error": str(e)}
    print(json.dumps({"pid": pid, "alive": bool(pid and pid_alive(pid)), "health": health}, indent=2))
    return 0


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("usage: gateway_supervisor.py start|stop|status", file=sys.stderr)
        return 64
    cmd = argv[1]
    if cmd == "start":
        return start()
    if cmd == "stop":
        return stop()
    if cmd == "status":
        return status()
    print(f"unknown subcommand: {cmd}", file=sys.stderr)
    return 64


if __name__ == "__main__":
    # gateway_client.py sits in the same dir; make it importable when this
    # script is invoked by path (the usual hook entrypoint pattern).
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    sys.exit(main(sys.argv))

File: scripts/gateway_supervisor.js
#!/usr/bin/env node
/**
 * Discover, start, and stop the memory-tencentdb Gateway Node.js sidecar.
 *
 * Usage:
 *   node scripts/gateway_supervisor.js start
 *   node scripts/gateway_supervisor.js stop
 *   node scripts/gateway_supervisor.js status
 */
"use strict";

const { spawn, execSync } = require("node:child_process");
const fs = require("node:fs");
const path = require("node:path");
const os = require("node:os");
const { GatewayClient } = require("./gateway_client.js");

const STATE_DIR = path.join(os.homedir(), ".memory-tencentdb");
const PID_FILE = path.join(STATE_DIR, "gateway.pid");
const LOG_DIR = path.join(STATE_DIR, "logs");
const STDOUT_LOG = path.join(LOG_DIR, "gateway.stdout.log");
const STDERR_LOG = path.join(LOG_DIR, "gateway.stderr.log");

const HEALTH_INTERVAL = 500;
const HEALTH_MAX_WAIT = 30000;

function discoverServerTs() {
  const home = os.homedir();
  const candidates = [
    path.join(home, ".memory-tencentdb", "tdai-memory-openclaw-plugin", "src", "gateway", "server.ts"),
    path.join(home, "tdai-memory-openclaw-plugin", "src", "gateway", "server.ts"),
    path.join(home, ".hermes", "plugins", "tdai-memory-openclaw-plugin", "src", "gateway", "server.ts"),
  ];
  for (const c of candidates) {
    if (fs.existsSync(c)) return c;
  }
  return null;
}

function resolveCommand() {
  const envCmd = (process.env.MEMORY_TENCENTDB_GATEWAY_CMD || "").trim();
  if (envCmd) return envCmd;
  const serverTs = discoverServerTs();
  if (serverTs) return `node --import tsx "${serverTs}"`;
  return null;
}

function readPid() {
  try { return parseInt(fs.readFileSync(PID_FILE, "utf-8").trim(), 10); } catch { return null; }
}

function pidAlive(pid) {
  try { process.kill(pid, 0); return true; } catch { return false; }
}

async function sleep(ms) { return new Promise((r) => setTimeout(r, ms)); }

async function start() {
  fs.mkdirSync(LOG_DIR, { recursive: true });

  const client = new GatewayClient();
  try {
    const h = await client.health(2000);
    console.log("gateway already healthy:", JSON.stringify(h));
    return 0;
  } catch {}

  const cmd = resolveCommand();
  if (!cmd) {
    console.error("ERROR: cannot locate Gateway. Set MEMORY_TENCENTDB_GATEWAY_CMD or run /memory-init.");
    return 2;
  }

  const stdoutFd = fs.openSync(STDOUT_LOG, "a");
  const stderrFd = fs.openSync(STDERR_LOG, "a");

  const parts = cmd.split(/\s+/);
  const proc = spawn(parts[0], parts.slice(1), {
    stdio: ["ignore", stdoutFd, stderrFd],
    detached: true,
    shell: process.platform === "win32",
  });
  proc.unref();
  fs.writeFileSync(PID_FILE, String(proc.pid), "utf-8");
  console.log(`spawned gateway pid=${proc.pid}, waiting for /health...`);

  const deadline = Date.now() + HEALTH_MAX_WAIT;
  while (Date.now() < deadline) {
    try {
      const h = await client.health(2000);
      if (h.status === "ok" || h.status === "degraded") {
        console.log("gateway ready:", JSON.stringify(h));
        fs.closeSync(stdoutFd);
        fs.closeSync(stderrFd);
        return 0;
      }
    } catch {}
    await sleep(HEALTH_INTERVAL);
  }

  fs.closeSync(stdoutFd);
  fs.closeSync(stderrFd);
  console.error("ERROR: gateway failed to become healthy within timeout");
  return 4;
}

function stop() {
  const pid = readPid();
  if (!pid) { console.log("no pid file; nothing to stop"); return 0; }
  if (!pidAlive(pid)) {
    console.log(`pid ${pid} already exited`);
    try { fs.unlinkSync(PID_FILE); } catch {}
    return 0;
  }
  try {
    if (process.platform === "win32") {
      execSync(`taskkill /PID ${pid} /T /F`, { stdio: "ignore" });
    } else {
      process.kill(pid, "SIGTERM");
    }
    console.log(`sent stop signal to pid ${pid}`);
  } catch (e) {
    console.error(`failed to signal pid ${pid}:`, e.message);
    return 1;
  }
  try { fs.unlinkSync(PID_FILE); } catch {}
  return 0;
}

async function status() {
  const pid = readPid();
  const client = new GatewayClient();
  let health;
  try { health = await client.health(2000); } catch (e) { health = { status: "down", error: e.message }; }
  console.log(JSON.stringify({ pid, alive: !!(pid && pidAlive(pid)), health }, null, 2));
  return 0;
}

async function main() {
  const cmd = process.argv[2];
  if (!cmd) { console.error("usage: gateway_supervisor.js start|stop|status"); process.exit(64); }
  if (cmd === "start") process.exit(await start());
  if (cmd === "stop") process.exit(stop());
  if (cmd === "status") process.exit(await status());
  console.error(`unknown subcommand: ${cmd}`);
  process.exit(64);
}

main();
